### PR TITLE
Get rid of package.json generation and yeoman generator

### PR DESCRIPTION
**Pull request by akosyakov**

### CODE REVIEW

**Comment by akosyakov** [comment]
Just to clarify: we do it to streamline the setup of Theia itself as well as the setup for extension developers and simplify the dynamic extension system.

An extension developer will be able to create package.json using standard tools, like `npm install`, list extension packages and add `@theia/cli` as a dev dependency:
https://github.com/theia-ide/theia/blob/44f764c9f195364ace851e88772c6302adef06b1/examples/browser/package.json#L1-L35
After installing such package he will use `theia` cli tool to manage an application, see npm scripts above. So such command like `npx theia browser --port=3001 --root-dir=/home/user/mydir` or via npm script `yarn start  --port=3001 --root-dir=/home/user/mydir` would be possible.

For the Theia setup, we won't need the package.json generation phase anymore that allows removing some build scripts, simplify root package.json, yarn workspaces.

**Comment by hexa00** [comment]
What is not supported by node 6 in this ?

**Comment by akosyakov** [comment]
@hexa00 async/await with pure js, bogus cluster module with node 6

**Comment by akosyakov** [comment]
LTS of node 8 should start in October: https://github.com/nodejs/Release#release-schedule1

**Comment by akosyakov** [comment]
I would like to merge if no objections and go on with the dynamic extension system.

File: dev-packages/cli/package.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    "electron-rebuild": "^1.5.11",
+    "file-loader": "^0.11.1",
+    "font-awesome-webpack": "0.0.5-beta.2",
+    "less": "^2.7.2",
+    "rimraf": "^2.6.1",
+    "source-map-loader": "^0.2.1",
+    "url-loader": "^0.5.8",
+    "webpack": "^2.2.1",
+    "yo": "^2.0.0",
+    "generator-theia": "^0.1.1"
```
Comment (svenefftinge) [inline dev-packages/cli/package.json]: `yargs`seems to be missing

**Comment by svenefftinge** [comment]
There are still some extension.package.json

**Comment by akosyakov** [comment]
@svenefftinge fixed your comments

File: package.json [hunk 2]
```diff
@@ -3,34 +3,53 @@
...
-    "build": "node scripts/lerna run build --scope \"@theia/*\" --stream",
-    "docs": "node scripts/lerna run docs --scope \"@theia/*\" --stream",
+    "prepare": "yarn run clean && yarn run rebuild:clean && yarn run build",
+    "clean": "node scripts/lerna run clean --parallel",
+    "build": "node scripts/lerna run build --stream",
+    "docs": "node scripts/lerna run docs --scope \"@theia/!(example-)*\" --stream",
     "test": "yarn run test:theia && yarn run test:electron && yarn run test:browser",
-    "test:theia": "node scripts/lerna run test --scope \"@theia/!(example-)*\" --parallel && yarn test:theia:slow",
-    "test:theia:slow": "node scripts/lerna run test:slow --scope \"@theia/terminal\" --parallel",
+    "test:theia": "node scripts/lerna run test --scope \"@theia/!(example-)*\" --parallel",
```
Comment (hexa00) [inline package.json]: why did you remove test:theia:slow?

Comment (akosyakov) [reply package.json]: do you run it?

Comment (akosyakov) [reply package.json]: I kept them for projects, e.g. `npx lerna run test:slow --scope @theia/terminal`

Comment (hexa00) [reply package.json]: Yes it was run by CI by default, I would like to keep it that way

Comment (akosyakov) [reply package.json]: it is still run by CI, each package has `test` script that runs `test:fast` + `test:slow`

Comment (hexa00) [reply package.json]: isn't that failing because many pacakge do not have slow tests ?

Comment (akosyakov) [reply package.json]: I would like to simplify package.json when it is possible and remove scripts which are just there without anybody calling them regularly

Comment (akosyakov) [reply package.json]: it is never failed if there are no tests, dummy files because of test coverage

**Comment by hexa00** [comment]
I get a bunch of warnings is that normal?
```
warning Pattern ["gulp-util@latest"] is trying to unpack in the same destination "/home/eantotr/.cache/yarn/v1/npm-gulp-util-3.0.8-0054e1e744502e27c04c187c3ecc505dd54bbb4f" as pattern ["gulp-util@^3.0.0","gulp-util@^3.0.1"]. This could result in a non deterministic behavior, skipping.
warning Pattern ["request@latest"] is trying to unpack in the same destination "/home/eantotr/.cache/yarn/v1/npm-request-2.82.0-2ba8a92cd7ac45660ea2b10a53ae67cd247516ea" as pattern ["request@^2.81.0","request@^2.72.0","request@2","request@^2.74.0","request@^2.45.0"]. This could result in a non deterministic behavior, skipping.
warning Pattern ["through@latest"] is trying to unpack in the same destination "/home/eantotr/.cache/yarn/v1/npm-through-2.3.8-0dd4c9ffaabc357960b1b724115d7e0e86a2e1f5" as pattern ["through@^2.3.6","through@^2.3.4","through@>=2.2.7 <3","through@2","through@~2.3.8","through@~2.3.4"]. This could result in a non deterministic behavior, skipping.
warning vscode-base-languageclient@0.0.1-alpha.3: The engine "vscode" appears to be invalid.
info fsevents@1.1.2: The platform "linux" is incompatible with this module.
info "fsevents@1.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning "font-awesome-webpack@0.0.5-beta.2" has incorrect peer dependency "font-awesome@>=4.3.0".
warning "url-loader@0.5.9" has incorrect peer dependency "file-loader@*".
warning "less-loader@2.2.3" has incorrect peer dependency "less@^2.3.1".
warning "ajv-keywords@1.5.1" has incorrect peer dependency "ajv@>=4.10.0".
warning The case-insensitive file /home/eantotr/sync/work/theia/node_modules/nyc/node_modules shouldn't be copied twice in one bulk copy
warning The case-insensitive file /home/eantotr/sync/work/theia/node_modules/nyc/node_modules/ansi-regex shouldn't be copied twice in one bulk copy
warning The case-insensitive file /home/eantotr/sync/work/theia/node_modules/nyc/node_modules/camelcase shouldn't be copied twice in one bulk copy


```

File: package.json (hunk 2 above)
Comment (hexa00) [reply package.json]: Yes the dummy files work for the test:fast pattern but not for the :slow

**Comment by akosyakov** [comment]
It does not seem to harm, but we should look into them for sure, could be done separately.

File: package.json (hunk 2 above)
Comment (akosyakov) [reply package.json]: it was the same before this PR should be tackled differently.

**Comment by hexa00** [comment]
```
theia/process:   6 passing (4s)
@theia/process:   1 failing
@theia/process:   1) RawProcess test pipe stderr stream:
@theia/process:      Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
@theia/process:   

@theia/editor: Warning: Could not find any test files matching pattern: ./src/**/*.slow-spec.ts
@theia/editor: No test files found
 
etc...
```
Build returns 0

**Comment by akosyakov** [comment]
@hexa00 Could you try it on master? Seems that some tests are bogus. Don't think that is related somehow to this PR.

File: package.json (hunk 2 above)
Comment (hexa00) [reply package.json]: No before this PR I explicitly selection that test:slow should be run on package/terminal only , otherwise the test run would fail

**Comment by akosyakov** [comment]
Nope, you are right, the timeout should fail the build, cannot find it in the log

**Comment by hexa00** [comment]
@akosyakov  ??? When I get an eror like that on maste the test run fails

**Comment by hexa00** [comment]
The @theia/editor: Warning: Could not find any test files matching pattern: ./src/**/*.slow-spec
also  failed the build before

File: package.json (hunk 2 above)
Comment (hexa00) [reply package.json]: See: the scope here:  `test:theia:slow": "node scripts/lerna run test:slow --scope \"@theia/terminal\" --parallel",`

**Comment by akosyakov** [comment]
@hexa00 you are right

**Comment by akosyakov** [comment]
@hexa00 What is with the test coverage if one runs nyc + mocha twice? Does it get merged? Or slow coverage overrides fast tests coverage?

**Comment by akosyakov** [comment]
@hexa00 I've merged `test:fast` and `test:slow` in single `test` and adjusted pattern to cover all `*.*spec.ts` files that there is no special handling for the terminal package and everything is covered
https://github.com/theia-ide/theia/blob/e36aa557bf687ece15a95b44c64e671b57445ea8/dev-packages/ext-scripts/package.json#L26-L27

**Comment by hexa00** [comment]
I think the coverage gets merged.
OK for the merged tests if we can separate it when need be.

**Comment by akosyakov** [comment]
@hexa00 yeah, it would be a bit messy with dummy spec files, we should look for the better way generally

**Comment by hexa00** [comment]
@akosyakov  Yes must be a way to treat warnings as warnings rather than errors

**Comment by hexa00** [comment]
Did you test the yarn.lock ?

**Comment by akosyakov** [comment]
@hexa00 I fixed the propagation of the error code broken tests should fail the build now

**Comment by hexa00** [comment]
Testing that now

**Comment by akosyakov** [comment]
@hexa00 nope, but should work, now lerna does not call install anymore, it is completely done by yarn

**Comment by hexa00** [comment]
Works fine , thanks! I need to increase that timeout

**Comment by hexa00** [comment]
Tested yarn.lock seems to work yeee!

**Comment by hexa00** [comment]
The log output is also so much easier to see!

**Comment by akosyakov** [comment]
I've merged theia-generator into the cli:
- there are no dependencies on yo and yeoman-generator anymore, it makes windows builds stable
- there are no yo configs, modules are discovered with `require.replace` after installing
- webpack configs are regenerated each time and added to git ignore
- moved all dev dependencies to the root package, new should be added there as well

Please review.
